# Exploring and clustering Toronto neighborhoods

In this Project we'll gather the toronto neighborhoods by scraping the Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, next, we'll clean the data and merge each postal code with its associated geospatial coordinates from the downloaded geospastial csv file.

For each neighborhood we'll get the closets 100 venues within a radios of 500m using the foursquare API, we'll then group the resuls by neighborhood and take the mean of the frequency of occurrence of each venue category, we'll put all this in a new DataFrame, which will be used later for the clustering stage where we used K-means to cluster the neighborhoods into 5 clusters.

finally we'll create one final DataFrame where we add the cluster associated to each neighborhood and then we'll print the map of toronto showing the different neighborhoods and their associated clusters.

- Let's Start by importing our packages

In [1]:
# Data Analysis and Manipulation
import numpy as np
import pandas as pd

import requests
import json

# Plotting Libraries and modules
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors
import seaborn as sns

# plotting Maps
import folium

# To retrieve Geospatial coordinates
from geopy.geocoders import Nominatim

# For Clustering
from sklearn.cluster import KMeans

plt.style.use('ggplot')
%matplotlib inline

## Gathering The Data

> In this step we'll scrape the PostalCode, Borough, and Neighborhoods from <a href= 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M,'>Wikipedia</a> using pandas read_html method to read the tables contained in the page, we'll set that na value is 'Not assigned'.

> As the web page contains multiple tables, we are interested only in the first table that contain the toronto data

In [2]:
## setting the link to the wikipedia webpage
URL = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

## reading the tables in the webpage and assign the first dataframe to scraped_df
scraped_df = pd.read_html(URL, na_values = 'Not assigned')[0]

## checking the DataFrame
scraped_df.head()

Postal Code           Borough               Neighborhood
0         M1A               NaN                        NaN
1         M2A               NaN                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

## Cleaning The Data

First We'll create a copy of our DataFrame, so changes does not affect the original one.

In [3]:
Toronto = scraped_df.copy()

Let's check our Data

In [4]:
## Checking for the datatype and null values of per column
Toronto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Postal Code   180 non-null    object
 1   Borough       103 non-null    object
 2   Neighborhood  103 non-null    object
dtypes: object(3)
memory usage: 4.3+ KB


In [5]:
## Checking the uniqueToronto Boroughs
Toronto['Borough'].unique()

array([nan, 'North York', 'Downtown Toronto', 'Etobicoke', 'Scarborough',
       'East York', 'York', 'East Toronto', 'West Toronto',
       'Central Toronto', 'Mississauga'], dtype=object)

> Because we will process only the cells that have an assigned borough, let's remove the Not assigned ones from our table

In [6]:
Toronto.dropna(subset=['Borough'], inplace= True)

Let's Check our Data again to confirm

In [7]:
Toronto.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 2 to 178
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Postal Code   103 non-null    object
 1   Borough       103 non-null    object
 2   Neighborhood  103 non-null    object
dtypes: object(3)
memory usage: 3.2+ KB


Is there any Duplicated Postal Code in our Data?

In [8]:
Toronto['Postal Code'].duplicated().sum()

0

How many Postal Code do we have in our dataset?

In [9]:
len(Toronto['Postal Code'].unique())

103

How many Boroughs are there?

In [10]:
len(Toronto.Borough.unique())

10

Let's print the first 5 rows to see how our new DatFrame looks like.

In [11]:
Toronto.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

What is The size of our Cleaned DataFrame?

In [12]:
Toronto.shape

(103, 3)

## Toronto Geographical Coordinates 

> to get the latitude and the longitude coordinates of each each postal code we'll download the https://cocl.us/Geospatial_data csv file and read it into a DataFrame.

In [13]:
geo_file = 'https://cocl.us/Geospatial_data'
geo_df = pd.read_csv(geo_file)

geo_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Now let's add the geographical coordinates of each postal code in the Toronto DataFrame

In [14]:
## Merging the two dataframes
Toronto = Toronto.merge(geo_df, how= 'inner', on= 'Postal Code')

Checking The Changes

In [15]:
Toronto.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

## The Analysis of Toronto Neighborhoods

First We'll use geopy library to get the latitude and longitude values of Toronto.

In [16]:
## Setting the Address variable
address = 'Toronto, Ontario'

## Getting the geographical coordinates
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

## Printing the coordinates
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Creating a map of Toronto with neighborhoods superimposed on top.

In [17]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(Toronto['Latitude'], Toronto['Longitude'],
                                           Toronto['Borough'], Toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

In [18]:
## Define Foursquare Credentials and Version

### Foursquare ID
CLIENT_ID = '***********************************************'

### Foursquare Secret
CLIENT_SECRET = '***********************************************'

### Foursquare API version
VERSION = '20200509'

Let's create a function to get the top 100 venues for each neighborhood in Toronto within a radius of 500 meters.

In [19]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, limit = 100):
    """
     function to get the top LIMIT venues for each neighborhood within a defined radius.
    """
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            limit)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now Let's execute our function

In [20]:
toronto_venues = getNearbyVenues(names=Toronto['Neighborhood'],
                                   latitudes=Toronto['Latitude'],
                                   longitudes=Toronto['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview
The Danforth West, Ri

**Let's check the size of the resulting dataframe**

In [21]:
## Checking the number of rows and columns
toronto_venues.shape

(2137, 7)

In [22]:
## Printing thefist 5 rows
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                        Venue  Venue Latitude  Venue Longitude  \
0             Brookbanks Park       43.751976       -79.332140   
1               Variety Store       43.751974       -79.333114   
2  TTC stop - 44 Valley Woods       43.755402       -79.333741   
3      Victoria Village Arena       43.723481       -79.315635   
4                 Tim Hortons       43.725517       -79.313103   

      Venue Category  
0               Park  
1  Food & Drink Shop  
2           Bus Stop  
3       Hockey Arena  
4        Coffee Shop

Let's check how many venues were returned for each neighborhood

In [23]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                              10   
Bathurst Manor, Wilson Heights, Downsview North                     19   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   22   
...                                                                ...   
Willowdale, Newtonbrook                                              1   
Woburn                                                               3   
Woodbine Heights                                                     9   
York Mills West                                                      4   
York Mills, Silver Hills                                             1   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                               10   
Bathurst Manor, Wilson Heights, Downsview North                      19   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Newtonbrook                                               1   
Woburn                                                                3   
Woodbine Heights                                                      9   
York Mills West                                                       4   
York Mills, Silver Hills                                              1   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                              10              10   
Bathurst Manor, Wilson Heights, Downsview North     19              19   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale, Newtonbrook                              1               1   
Woburn                                               3               3   
Woodbine Heights                                     9               9   
York Mills West                                      4               4   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                        10   
Bathurst Manor, Wilson Heights, Downsview North               19   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale, Newtonbrook                                        1   
Woburn                                                         3   
Woodbine Heights                                               9   
York Mills West                                                4   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                       10 

> Since we have 103 rows in our Toronto DataFrame, and we get the results for 96 Neighborhoods
> Let's check which Neighborhoods did not get any results

In [24]:
Neighborhoods = Toronto['Neighborhood'].values
venues_Neighborhoods = toronto_venues['Neighborhood'].values


for Neighborhood in Neighborhoods:
    
    if Neighborhood not in venues_Neighborhoods:
        print(Neighborhood)

Islington Avenue
Upper Rouge


Let's Drop thoese Neighborhoods from our dataset

In [25]:
## Retrieving the indices of "Islington Avenue", "Upper Rouge"
ind_to_drop = Toronto.query('Neighborhood in ("Islington Avenue", "Upper Rouge")').index

## Dropping "Islington Avenue", "Upper Rouge" from the DataFrame
Toronto.drop(ind_to_drop, inplace= True)

Let's find out how many unique categories can be curated from all the returned venues

In [26]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 266 uniques categories.


## Analyzing Each Neighborhood

In [27]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Trail  Train Station  \
0                 0                    0  ...      0              0   
1                 0                    0  ...      0              0   
2                 0                    0  ...      0              0   
3                 0                    0  ...      0              0   
4                 0                    0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 266 columns]

Let's examine the new dataframe size.

In [28]:
toronto_onehot.shape

(2137, 266)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [29]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Agincourt          0.0   
1                            Alderwood, Long Branch          0.0   
2   Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                   Bayview Village          0.0   
4                 Bedford Park, Lawrence Manor East          0.0   
..                                              ...          ...   
91                          Willowdale, Newtonbrook          0.0   
92                                           Woburn          0.0   
93                                 Woodbine Heights          0.0   
94                                  York Mills West          0.0   
95                         York Mills, Silver Hills          0.0   

    Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                 0.0                0.0      0.0                 0.0   
1                 0.0                0.0      0.0                 0.0   
2                 0.0                0.0      0.0                 0.0   
3                 0.0                0.0      0.0                 0.0   
4                 0.0                0.0      0.0                 0.0   
..                ...                ...      ...                 ...   
91                0.0                0.0      0.0                 0.0   
92                0.0                0.0      0.0                 0.0   
93                0.0                0.0      0.0                 0.0   
94                0.0                0.0      0.0                 0.0   
95                0.0                0.0      0.0                 0.0   

    Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0            0.0             0.0              0.0               0.0  ...   
1            0.0             0.0              0.0               0.0  ...   
2            0.0             0.0              0.0               0.0  ...   
3            0.0             0.0              0.0               0.0  ...   
4            0.0             0.0              0.0               0.0  ...   
..           ...             ...              ...               ...  ...   
91           0.0             0.0              0.0               0.0  ...   
92           0.0             0.0              0.0               0.0  ...   
93           0.0             0.0              0.0               0.0  ...   
94           0.0             0.0              0.0               0.0  ...   
95           0.0             0.0              0.0               0.0  ...   

    Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0     0.0            0.0                            0.0               0.0   
1     0.0            0.0                            0.0               0.0   
2     0.0            0.0                            0.0               0.0   
3     0.0            0.0                            0.0               0.0   
4     0.0            0.0                            0.0               0.0   
..    ...            ...                            ...               ...   
91    0.0            0.0                            0.0               0.0   
92    0.0            0.0                            0.0               0.0   
93    0.0            0.0                            0.0               0.0   
94    0.0            0.0                            0.0               0.0   
95    0.0            0.0                            0.0               0.0   

    Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0      0.000000                    0.0              0.0       0.0   
1      0.000000                    0.0              0.0       0.0   
2      0.052632                    0.0              0.0       0.0   
3      0.000000                    0.0              0.0       0.0   
4      0.000000                    0.0              0.0       0.0   
..          ...                    ...              ...       ...   
91     0.000000         

Let's confirm the new size

In [30]:
toronto_grouped.shape

(96, 266)

Let's print each neighborhood along with the top 5 most common venues

In [31]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0                     Lounge  0.25
1               Skating Rink  0.25
2  Latin American Restaurant  0.25
3             Breakfast Spot  0.25
4   Mediterranean Restaurant  0.00


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place   0.2
1             Gym   0.1
2  Sandwich Place   0.1
3             Pub   0.1
4            Pool   0.1


----Bathurst Manor, Wilson Heights, Downsview North----
                       venue  freq
0                       Bank  0.11
1                Coffee Shop  0.11
2        Fried Chicken Joint  0.05
3  Middle Eastern Restaurant  0.05
4                Gas Station  0.05


----Bayview Village----
                 venue  freq
0                 Café  0.25
1                 Bank  0.25
2  Japanese Restaurant  0.25
3   Chinese Restaurant  0.25
4          Yoga Studio  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0  Italian Restaurant  0.09
1         Coffee Shop  0.09

         venue  freq
0         Café  0.13
1     Pharmacy  0.07
2       Bakery  0.07
3  Pizza Place  0.07
4  Coffee Shop  0.07


----North Park, Maple Leaf Park, Upwood Park----
                        venue  freq
0                        Park  0.33
1                      Bakery  0.33
2  Construction & Landscaping  0.33
3                 Men's Store  0.00
4           Mobile Phone Shop  0.00


----North Toronto West----
                venue  freq
0      Clothing Store  0.13
1         Coffee Shop  0.09
2         Yoga Studio  0.04
3  Miscellaneous Shop  0.04
4        Dessert Shop  0.04


----Northwest----
                             venue  freq
0              Rental Car Location   0.5
1                        Drugstore   0.5
2  Molecular Gastronomy Restaurant   0.0
3                    Luggage Store   0.0
4                           Market   0.0


----Northwood Park, York University----
                  venue  freq
0    Falafel Restaurant   0.2
1        Massage Studio   0.2
2           

Let's put that into a **pandas** dataframe

- First, let's write a function to sort the venues in descending order.

In [32]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

- Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [56]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village         Chinese Restaurant   
4                Bedford Park, Lawrence Manor East                Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot                Lounge          Skating Rink   
1          Skating Rink        Sandwich Place              Pharmacy   
2                  Bank      Sushi Restaurant           Bridal Shop   
3                  Café                  Bank   Japanese Restaurant   
4            Restaurant        Sandwich Place    Italian Restaurant   

       5th Most Common Venue    6th Most Common Venue  \
0       Colombian Restaurant  Comfort Food Restaurant   
1                       Pool                      Pub   
2  Middle Eastern Restaurant            Deli / Bodega   
3              Women's Store             Dessert Shop   
4              Grocery Store         Sushi Restaurant   

     7th Most Common Venue        8th Most Common Venue 9th Most Common Venue  \
0        Electronics Store  Eastern European Restaurant             Drugstore   
1             Dance Studio                  Coffee Shop                   Gym   
2               Restaurant                  Pizza Place              Pharmacy   
3       Dim Sum Restaurant                        Diner        Discount Store   
4  Comfort Food Restaurant                     Pharmacy           Pizza Place   

  10th Most Common Venue  
0     College Rec Center  
1            Gas Station  
2    Fried Chicken Joint  
3    Distribution Center  
4                   Café

## Cluster Neighborhoods

We'll Run *k*-means to cluster the neighborhoods into 5 clusters.

In [57]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(init="k-means++", n_clusters=kclusters, n_init=15, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 3, 0, 0, 1, 3, 0, 2, 0, 3, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 3,
       0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0,
       3, 0, 0, 4, 0, 0, 3, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [58]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = Toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656               3                  Park   
1  43.725882 -79.315572               0     French Restaurant   
2  43.654260 -79.360636               0           Coffee Shop   
3  43.718518 -79.464763               0        Clothing Store   
4  43.662301 -79.389494               0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0     Food & Drink Shop              Bus Stop     College Rec Center   
1           Coffee Shop          Hockey Arena  Portuguese Restaurant   
2                  Park                   Pub                 Bakery   
3     Accessories Store           Event Space             Shoe Store   
4      Sushi Restaurant          Burger Joint                    Bar   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0           Dance Studio  Eastern European Restaurant             Drugstore   
1     Dim Sum Restaurant                  Curling Ice          Dance Studio   
2                   Café                      Theater            Restaurant   
3  Vietnamese Restaurant           Miscellaneous Shop           Coffee Shop   
4               Beer Bar           Italian Restaurant             Juice Bar   

  8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0            Donut Shop      Doner Restaurant                 Dog Run  
1         Deli / Bodega      Department Store            Dessert Shop  
2        Breakfast Spot           Yoga Studio             Event Space  
3              Boutique   Arts & Crafts Store  Furniture / Home Store  
4        Sandwich Place   Distribution Center              Restaurant

Finally, let's visualize the resulting clusters

In [59]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examining The Clusters

> Now, we can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories.

#### Cluster 1

In [60]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1          North York               0           French Restaurant   
2    Downtown Toronto               0                 Coffee Shop   
3          North York               0              Clothing Store   
4    Downtown Toronto               0                 Coffee Shop   
7          North York               0                 Coffee Shop   
..                ...             ...                         ...   
97   Downtown Toronto               0                 Coffee Shop   
99   Downtown Toronto               0                 Coffee Shop   
100      East Toronto               0                 Yoga Studio   
101         Etobicoke               0  Construction & Landscaping   
102         Etobicoke               0               Grocery Store   

    2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1             Coffee Shop          Hockey Arena  Portuguese Restaurant   
2                    Park                   Pub                 Bakery   
3       Accessories Store           Event Space             Shoe Store   
4        Sushi Restaurant          Burger Joint                    Bar   
7        Asian Restaurant            Beer Store    Japanese Restaurant   
..                    ...                   ...                    ...   
97                   Café                 Hotel                    Gym   
99       Sushi Restaurant   Japanese Restaurant             Restaurant   
100         Auto Workshop            Comic Shop                   Park   
101        Baseball Field        Breakfast Spot           Dessert Shop   
102           Flower Shop           Wings Joint                    Gym   

     5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
1       Dim Sum Restaurant           Curling Ice              Dance Studio   
2                     Café               Theater                Restaurant   
3    Vietnamese Restaurant    Miscellaneous Shop               Coffee Shop   
4                 Beer Bar    Italian Restaurant                 Juice Bar   
7                      Gym            Restaurant                 Bike Shop   
..                     ...                   ...                       ...   
97     Japanese Restaurant            Restaurant        Seafood Restaurant   
99                     Pub           Men's Store  Mediterranean Restaurant   
100            Pizza Place               Butcher          Recording Studio   
101     Dim Sum Restaurant                 Diner            Discount Store   
102                 Bakery        Sandwich Place      Fast Food Restaurant   

    8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
1           Deli / Bodega      Department Store            Dessert Shop  
2          Breakfast Spot           Yoga Studio             Event Space  
3                Boutique   Arts & Crafts Store  Furniture / Home Store  
4          Sandwich Place   Distribution Center              Restaurant  
7      Italian Restaurant        Discount Store     Sporting Goods Shop  
..                    ...                   ...                     ...  
97            Salad Place            Steakhouse        Asian Restaurant  
99             Smoke Shop                 Hotel                 Gay Bar  
100         Burrito Place            Restaurant                 Brewery  
101   Distribution Center               Dog Run           Women's Store  
102        Hardware Store       Supplement Shop       Convenience Store  

[87 rows x 12 columns]

#### Cluster 2

In [61]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough               1  Fast Food Restaurant         Women's Store   

  3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
6           Curling Ice  Eastern European Restaurant             Drugstore   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6            Donut Shop      Doner Restaurant               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
6   Distribution Center         Discount Store

#### Cluster 3

In [62]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
32      Scarborough               2     Convenience Store   
83  Central Toronto               2            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32            Playground         Women's Store        Discount Store   
83         Women's Store        Discount Store          Dance Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32         Deli / Bodega      Department Store          Dessert Shop   
83         Deli / Bodega      Department Store          Dessert Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
32    Dim Sum Restaurant                 Diner    Distribution Center  
83    Dim Sum Restaurant                 Diner    Distribution Center

#### Cluster 4

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York               3                  Park   
21              York               3                  Park   
35         East York               3                  Park   
49        North York               3                  Park   
61   Central Toronto               3                  Park   
64              York               3                  Park   
66        North York               3                  Park   
85       Scarborough               3                  Park   
91  Downtown Toronto               3                  Park   
98         Etobicoke               3                  Park   

         2nd Most Common Venue 3rd Most Common Venue  \
0            Food & Drink Shop              Bus Stop   
21                        Pool         Women's Store   
35                 Coffee Shop     Convenience Store   
49  Construction & Landscaping                Bakery   
61                 Swim School              Bus Line   
64                 Curling Ice     Electronics Store   
66                        Bank     Convenience Store   
85                      Bakery            Playground   
91                       Trail            Playground   
98                       River          Cupcake Shop   

          4th Most Common Venue 5th Most Common Venue  \
0            College Rec Center          Dance Studio   
21            Afghan Restaurant  Ethiopian Restaurant   
35               Discount Store         Deli / Bodega   
49             Department Store          Dessert Shop   
61                    Drugstore            Donut Shop   
64  Eastern European Restaurant             Drugstore   
66               Discount Store      Department Store   
85               Discount Store         Deli / Bodega   
91               Discount Store          Dance Studio   
98                    Drugstore            Donut Shop   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0   Eastern European Restaurant             Drugstore            Donut Shop   
21  Eastern European Restaurant             Drugstore            Donut Shop   
35             Department Store          Dessert Shop    Dim Sum Restaurant   
49           Dim Sum Restaurant                 Diner        Discount Store   
61             Doner Restaurant               Dog Run   Distribution Center   
64                   Donut Shop      Doner Restaurant               Dog Run   
66                 Dessert Shop    Dim Sum Restaurant                 Diner   
85             Department Store          Dessert Shop    Dim Sum Restaurant   
91                Deli / Bodega      Department Store          Dessert Shop   
98             Doner Restaurant               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
0       Doner Restaurant                Dog Run  
21      Doner Restaurant                Dog Run  
35                 Diner    Distribution Center  
49   Distribution Center           Dance Studio  
61  Ethiopian Restaurant                  Diner  
64   Distribution Center         Discount Store  
66   Distribution Center           Dance Studio  
85                 Diner    Distribution Center  
91    Dim Sum Restaurant                  Diner  
98        Discount Store                  Diner

#### Cluster 5

In [64]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
52  North York               4          Home Service         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
52        Discount Store         Deli / Bodega      Department Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
52          Dessert Shop    Dim Sum Restaurant                 Diner   

   9th Most Common Venue 10th Most Common Venue  
52   Distribution Center            Curling Ice

## Result

> Although we initialised the k-means with 5 K, it seems like there is two main distinct clusters in toronto, where the main distinctive venue is the park.

> - **Cluster 1** includes most neighborhoods, where we mainly find restaurants, coffe shops, and clothes / accessories shops.
>  

> - The second one **Cluster 4**  where the park is its main common venue.